### PR TITLE
PERF-2676 TPCC fails for MongoDB on Atlas

### PR DESCRIPTION
**Pull request by apodelko**
A short-term fix for py-tpcc to allow it to run on Atlas. See details on PERF-2676.

### CODE REVIEW

**Review by dalyd — APPROVED** [review]

File: pytpcc/drivers/mongodbdriver.py [hunk 1]
```diff
@@ -1149,6 +1149,6 @@ def get_server_status(self):
     def save_result(self, result_doc):
         self.result_doc.update(result_doc)
         self.result_doc['after']=self.get_server_status()
-        self.client.test.results.save(self.result_doc)
+        self.client.test.results.save(self.result_doc, check_keys=False)
```
Comment (dalyd) [inline pytpcc/drivers/mongodbdriver.py:1152]: It looks like save is deprecated. should this be an insert_one or replace_one?

**Comment by apodelko** [comment]
Here is the variant with insert_one. Here is [the patch](https://spruce.mongodb.com/version/6199939b0305b902881d8896/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC). But we need new Pymongo 3.12.1 - otherwise we get the same error in Atlas with MongoDb 5.0 (we have currently 3.11.1 in workload_setup.common.yml). It appears that we officially need 3.12.1 to support MongoDB 5.0 - so it is probably the right thing to do. The suggestion is to use insert_one as advised by the Python driver team and update Pymongo to 3.12.1 (for which I will create another PR in dsi if we agree on this approach). Any thoughts?

File: pytpcc/drivers/mongodbdriver.py [hunk 2]
```diff
@@ -1149,6 +1149,6 @@ def get_server_status(self):
     def save_result(self, result_doc):
         self.result_doc.update(result_doc)
         self.result_doc['after']=self.get_server_status()
-        self.client.test.results.save(self.result_doc)
+        self.client.test.results.insert_one(self.result_doc)
```
Comment (jimoleary) [inline pytpcc/drivers/mongodbdriver.py:1152]: I don't think this is correct, save is either an insert or a replace. The logic above only mimics the first part.

Should this be an upsert (see [here](https://github.com/mongodb/mongo-python-driver/blob/24cc4c42bf6476c25f90cfcc7e10e63e359465b2/doc/migrate-to-pymongo4.rst#collectionsave-is-removed))

**Review by jimoleary — CHANGES_REQUESTED** [review]

File: pytpcc/drivers/mongodbdriver.py (hunk 2 above)
Comment (apodelko) [reply pytpcc/drivers/mongodbdriver.py:1152]: @jimoleary That is correct, it does formally change the logic. The literal replacement would be  with the whole code you linked above. However as far as I was looking into what is going on here, it appears that only insert happens here - there is nothing to update. Unless I missed something here, why we should process update_one logic here if it never happens?

Comment (jimoleary) [reply pytpcc/drivers/mongodbdriver.py:1152]: I'm not as familiar with this code so I'll defer to you. But can we add a comment to say that this is the case?

**Comment by apodelko** [comment]
Actually we have pymongo upgrade in another [PR ](https://github.com/10gen/dsi/pull/985)

File: pytpcc/drivers/mongodbdriver.py (hunk 2 above)
Comment (apodelko) [reply pytpcc/drivers/mongodbdriver.py:1153]: added a comment - let me know if it helps

Comment (dalyd) [reply pytpcc/drivers/mongodbdriver.py:1153]: Comment looks good. I don't believe we ever set an explicit _id field in result_doc, so this should always be an insert.

Comment (dalyd) [reply pytpcc/drivers/mongodbdriver.py:1153]: I'm also not a fan of saving the results back into the database itself, but that's a separate issue.

**Review by dalyd — APPROVED** [review]
Update to insert_one sounds good to me. LGTM.

**Pull request merged**